packages/trace-viewer/src/ui/consoleTab.tsx

```diff
@@ -139,3 +148,75 @@ function iconClass(message: channels.ConsoleMessageInitializer): string {
 function stdioClass(isError: boolean): string {
   return isError ? 'error' : 'blank';
 }
+
+function format(args: { preview: string, value: any }[]): JSX.Element[] {
+  if (args.length === 1)
+    return [<span>{args[0].preview}</span>];
+  const hasMessageFormat = typeof args[0].value === 'string' && args[0].value.includes('%');
+  const messageFormat = hasMessageFormat ? args[0].value as string : '';
+  const tail = hasMessageFormat ? args.slice(1) : args;
+  let argIndex = 0;
+
+  const regex = /%([%sdifoOc])/g;
+  let match;
+  const formatted: JSX.Element[] = [];
+  let tokens: JSX.Element[] = [];
+  formatted.push(<span>{tokens}</span>);
+  let formatIndex = 0;
+  while ((match = regex.exec(messageFormat)) !== null) {
+    const text = messageFormat.substring(formatIndex, match.index);
+    tokens.push(<span>{text}</span>);
+    formatIndex = match.index + 2;
+    const specifier = match[0][1];
+    if (specifier === '%') {
+      tokens.push(<span>%</span>);
+    } else if (specifier === 's' || specifier === 'o' || specifier === 'O' || specifier === 'd' || specifier === 'i' || specifier === 'f') {
+      const value = tail[argIndex++];
+      const styleObject: any = {};
+      if (typeof value?.value !== 'string')
+        styleObject['color'] = 'var(--vscode-debugTokenExpression-number)';
+      tokens.push(<span style={styleObject}>{value?.preview || ''}</span>);
+    } else if (specifier === 'c') {
+      tokens = [];
+      const format = tail[argIndex++];
+      const styleObject = format ? parseCSSStyle(format.preview) : {};
+      formatted.push(<span style={styleObject}>{tokens}</span>);
+    }
+  }
+  if (formatIndex < messageFormat.length)
+    tokens.push(<span>{messageFormat.substring(formatIndex)}</span>);
+  for (; argIndex < tail.length; argIndex++) {
+    const value = tail[argIndex];
+    const styleObject: any = {};
+    if (tokens.length)
+      tokens.push(<span> </span>);
+    if (typeof value?.value !== 'string')
+      styleObject['color'] = 'var(--vscode-debugTokenExpression-number)';
+    tokens.push(<span style={styleObject}>{value?.preview || ''}</span>);
+  }
+  return formatted;
+}
+
+function parseCSSStyle(cssFormat: string): Record<string, string | number> {
+  try {
+    const styleObject: Record<string, string | number> = {};
+    const cssText = cssFormat.replace(/;$/, '').replace(/: /g, ':').replace(/; /g, ';');
```

Comment:
Can we just trim the values after splitting?

Reply:
I'll follow up